lib/pdn/pdn.c

```diff
@@ -0,0 +1,768 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <pdn.h>
+#include <at_cmd.h>
+#include <at_notif.h>
+
+#include <init.h>
+#include <errno.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <kernel.h>
+
+/** Maximum buffer size for AT commands.
+ * Max size is defined to contain commands and responses.
+ * Longest AT command is CONNECT:
+ *   AT+CGDCONT=CID,"FAMILY","APN"
+ *   where:
+ *     CID     up to 2 octets
+ *     FAMILY  up to 6 octets
+ *     APN     up to 100 octets
+ *   => 11+2+1+8+1+102+2 = 127 octets
+ *
+ * Longest received response is for PDN_STATE_GET
+ *   "+CGACT: 0,1\r\n
+ *    ...
+ *    +CGACT: 11,1\r\n
+ *    OK\r\n"
+ *   => 10*(10+2)+2*(11+2)+2+2 = 150 octets
+ */
+#define PDN_AT_BUFFER_SIZE (192)
+
+#define PDN_ID_INVALID (-1)
+#define PDN_CID_INVALID (-1)
+
+#define PDN_IDX_TO_HANDLE(_idx) ((int)_idx + CONFIG_PDN_NUM_CONTEXTS)
+#define PDN_HANDLE_TO_IDX(_handle) ((int)_handle - CONFIG_PDN_NUM_CONTEXTS)
+
+/** @brief PDN states
+ * Internal definitions to track PDN state
+ */
+enum pdn_conn_state {
+	/** PDN is free. */
+	PDN_STATE_FREE = 0,
+	/** PDN is allocated. */
+	PDN_STATE_ALLOCATED,
+	/** PDN is idle/unused. */
+	PDN_STATE_IDLE,
+	/** PDN is connected to an APN and activated. */
+	PDN_STATE_ACTIVE
+};
+
+/** @brief AT commands
+ * Internal definitions for supported AT command
+ */
+enum pdn_at_cmd {
+	CMD_CONTEXT_CREATE = 0,
+	CMD_APN_CONNECT,
+	CMD_PDN_ACTIVATE,
+	CMD_PDN_ID_GET,
+	CMD_PDN_STATE_GET,
+	CMD_PDN_DEACTIVATE,
+	CMD_APN_DISCONNECT,
+	CMD_TIME_GET,
+	CMD_MAX
+};
+
+/** @brief AT command responses
+ * Internal definitions for supported AT command responses
+ */
+enum pdn_at_evt {
+	EVT_CONTEXT_CREATE = 0,
+	EVT_PDN_ID_GET,
+	EVT_PDN_STATE_GET,
+	EVT_TIME_GET,
+	EVT_MAX
+};
+
+/** Data type for PDN connection. */
+struct pdn_connection {
+	/* Client defined callback when PDN related nofication is received */
+	pdn_notif_handler_t callback;
+	/* Last connect / disconnect time */
+	struct pdn_time event_time;
+	/* PDN ID. */
+	s8_t pdn_id;
+	/* Indicates the connection state #pdn_conn_state of the PDN. */
+	u8_t state;
+	/* Supported address family. @ref #pdn_inet_type */
+	u8_t family;
```

Comment:
Why not `enum pdn_event`?

Reply:
to save memory, no need to use int type to store just couple of bits.